ampd/src/mvx/verifier.rs

```diff
@@ -0,0 +1,620 @@
+        if event.identifier != CONTRACT_CALL_IDENTIFIER {
+            return Ok(false);
+        }
+
+        let topics = event.topics.as_ref().ok_or(Error::PropertyEmpty)?;
+
+        let event_name = topics.get(0).ok_or(Error::PropertyEmpty)?;
+        let event_name = STANDARD.decode(event_name)?;
+        if event_name.as_slice() != CONTRACT_CALL_EVENT.as_bytes() {
+            return Ok(false);
+        }
+
+        let sender = topics.get(1).ok_or(Error::PropertyEmpty)?;
+        let sender = STANDARD.decode(sender)?;
+        if sender.len() != 32 || &sender[0..32] != &self.source_address.to_bytes() {
+            return Ok(false);
+        }
+
+        let destination_chain = topics.get(2).ok_or(Error::PropertyEmpty)?;
+        let destination_chain = STANDARD.decode(destination_chain)?;
+        let destination_chain = String::from_utf8(destination_chain)?;
+        if destination_chain != self.destination_chain.to_string() {
+            return Ok(false);
+        }
+
+        let destination_address = topics.get(3).ok_or(Error::PropertyEmpty)?;
+        let destination_address = STANDARD.decode(destination_address)?;
+        let destination_address = String::from_utf8(destination_address)?;
+        if destination_address != self.destination_address {
+            return Ok(false);
+        }
+
+        let payload_hash = topics.get(4).ok_or(Error::PropertyEmpty)?;
+        let payload_hash = STANDARD.decode(payload_hash)?;
+        if payload_hash.len() != 32
+            || Hash::from_slice(payload_hash.as_slice()) != self.payload_hash
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+impl VerifierSetConfirmation {
+    fn eq_event(&self, event: &Events) -> Result<bool, Box<dyn std::error::Error>> {
+        if event.identifier != ROTATE_SIGNERS_IDENTIFIER {
+            return Ok(false);
+        }
+
+        let topics = event.topics.as_ref().ok_or(Error::PropertyEmpty)?;
+
+        let event_name = topics.get(0).ok_or(Error::PropertyEmpty)?;
+        let event_name = STANDARD.decode(event_name)?;
+        if event_name.as_slice() != SIGNERS_ROTATED_EVENT.as_bytes() {
+            return Ok(false);
+        }
+
+        let signers_hash = topics.get(2).ok_or(Error::PropertyEmpty)?;
+        let signers_hash = STANDARD.decode(signers_hash)?;
+
+        let weighted_signers = WeightedSigners::from(&self.verifier_set);
+
+        if signers_hash.len() != 32 || signers_hash.as_slice() != weighted_signers.hash().as_slice()
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+fn find_event<'a>(
+    transaction: &'a TransactionOnNetwork,
+    gateway_address: &Address,
+    log_index: usize,
+    identifier: &str,
+    needed_event_name: &[u8],
+) -> Option<&'a Events> {
+    // We only support log index being 0, so one supported event per transaction because of MultiversX limitations
+    if log_index != 0 {
+        return None;
+    }

```

Comment:
Can you expand more on why this is the case? Not in the code necessarily, but just in a comment. Will this always be the case? If so, why not just use the transaction hash by itself as the message id?

Reply:
Will add a comment. I want to keep the ability to modify this in the future, maybe with an activation epoch flag (eg: after epoch X, we consider only messages with log index, before that we consider only log index 0).

I wanted to also leave the message id like it is since then I won't have to add a custom message id type in the library and it is easier this way.